### PR TITLE
Feature/missions

### PR DESCRIPTION
**Pull request by ylvaselling**
To be used with the `feature/missions` gui

Continued the work Micah did with the missions topic. 

* Added GUI with timeline for phases, instrument activities and milestones
* Added Rosetta mission file
* Modified OSIRIS-REx mission file

To try: start a profile with a mission.asset file, e.g. New Horizons, Rosetta, or OSIRIS-Rex

### CODE REVIEW

**Comment by alexanderbock** [comment]
When selecting "Current Phase" in the top, the selection should probably update when we move from phase to phase?  Example in the picture:  I was past the current phase, clicked on "Current Phase" and then moved backwards in time
image

Would have expected to see:
image

**Comment by alexanderbock** [comment]
I'm sure it has been there before, but only seeing it now;  when the mouse cursor is over the Mission panel, the bottom bar also highlights:
image
vs
image


*super* minor detail of course

**Comment by alexanderbock** [comment]
In the New Horizons profile when hovering close to the "current time indicator":
image

and the current phase and tooltip flickers:
Uploading 2023-04-12 23-58-43.mp4…

File: data/assets/scene/solarsystem/missions/newhorizons/mission.asset [hunk 1]
```diff
@@ -120,3 +120,11 @@ return {
...
 }
 
+asset.onInitialize(function()
+  openspace.loadMission(Mission)
+ end)
+ 
+ asset.onDeinitialize(function()
+  openspace.unloadMission(Mission.Name)
+ end)
+
```
Comment (alexanderbock) [inline data/assets/scene/solarsystem/missions/newhorizons/mission.asset:130]: Missing empty line at the end of the file

File: data/assets/scene/solarsystem/missions/osirisrex/actions.asset [hunk 2]
```diff
@@ -83,6 +83,59 @@ local toggle_sun_marker = {
...
+local aimAtBennu = {
+  Identifier = "os.missions.osirisrex.aimAtBennu",
+  Name = "Look at Bennu from OSIRIS-REx",
+  Command = [[
+      openspace.setPropertyValueSingle("NavigationHandler.OrbitalNavigator.RetargetAnchor", nil);
+      openspace.setPropertyValueSingle("NavigationHandler.OrbitalNavigator.Anchor", 'OsirisRex');
+      openspace.setPropertyValueSingle("NavigationHandler.OrbitalNavigator.Aim", 'Earth');
+      openspace.setPropertyValueSingle("NavigationHandler.OrbitalNavigator.RetargetAim", nil);
+      openspace.setPropertyValueSingle("NavigationHandler.OrbitalNavigator.Anchor", 'OsirisRex');
+      openspace.setPropertyValueSingle("NavigationHandler.OrbitalNavigator.Aim", 'BennuBarycenter');
```
Comment (alexanderbock) [inline data/assets/scene/solarsystem/missions/osirisrex/actions.asset:95]: Do we need to do the aiming towards Earth and _then_ the aiming towards Bennu?

File: data/assets/scene/solarsystem/missions/osirisrex/osirisrex.asset [hunk 3]
```diff
@@ -4,16 +4,6 @@ asset.require("./model")
...
-local missionName
-
-asset.onInitialize(function()
-  missionName = openspace.loadMission(mission)
-end)
-
-asset.onDeinitialize(function()
-  openspace.unloadMission(missionName)
-end)
+asset.require('./mission')
```
Comment (alexanderbock) [inline data/assets/scene/solarsystem/missions/osirisrex/osirisrex.asset:9]: Missing empty line at the end of the file

File: data/assets/scene/solarsystem/missions/perseverance/mission.asset [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      },
+      {
+        Name = "Drive Test",
+        TimeRange = { Start = "2021 MAR 07" },
+        Media = { Image = "https://upload.wikimedia.org/wikipedia/commons/2/21/PIA23729-MarsPerseveranceRover-DriveTest-20210307.jpg" }
+      },
+      {
+        Name = "Ingenuity Helicopter launch",
+        TimeRange = { Start = "2021 APR 04" },
+        Media = { Image = "https://upload.wikimedia.org/wikipedia/commons/e/e1/Ingenuity_helicopter_first_colour_image.jpg" }
```
Comment (alexanderbock) [inline data/assets/scene/solarsystem/missions/perseverance/mission.asset:23]: -> http://data.openspaceproject.com/missions/perseverance/Ingenuity_helicopter_first_colour_image.jpg

File: data/assets/scene/solarsystem/missions/perseverance/mission.asset [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      },
+      {
+        Name = "First Study Target",
+        TimeRange = { Start = "2021 FEB 22" },
+        Media = { Image = "https://upload.wikimedia.org/wikipedia/commons/6/69/PIA24484-MarsPerseveranceRover-FirstStudyTarget-20210222.jpg" }
+      },
+      {
+        Name = "Drive Test",
+        TimeRange = { Start = "2021 MAR 07" },
+        Media = { Image = "https://upload.wikimedia.org/wikipedia/commons/2/21/PIA23729-MarsPerseveranceRover-DriveTest-20210307.jpg" }
```
Comment (alexanderbock) [inline data/assets/scene/solarsystem/missions/perseverance/mission.asset:18]: -> http://data.openspaceproject.com/missions/perseverance/PIA23729-MarsPerseveranceRover-DriveTest-20210307.jpg

File: data/assets/scene/solarsystem/missions/perseverance/mission.asset [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    Phases = {
+      {
+        Name = "First Color Photo",
+        TimeRange = { Start = "2021 FEB 18" },
+        Media = { Image = "https://upload.wikimedia.org/wikipedia/commons/d/d6/Perseverance%27s_First_Full-Color_Look_at_Mars.png" }
+      },
+      {
+        Name = "First Study Target",
+        TimeRange = { Start = "2021 FEB 22" },
+        Media = { Image = "https://upload.wikimedia.org/wikipedia/commons/6/69/PIA24484-MarsPerseveranceRover-FirstStudyTarget-20210222.jpg" }
```
Comment (alexanderbock) [inline data/assets/scene/solarsystem/missions/perseverance/mission.asset:13]: -> http://data.openspaceproject.com/missions/perseverance/PIA24484-MarsPerseveranceRover-FirstStudyTarget-20210222.jpg

File: data/assets/scene/solarsystem/missions/perseverance/mission.asset [hunk 7]
```diff
@@ -0,0 +1,40 @@
+-- @TODO (Images should be moved to the data server)
```
Comment (alexanderbock) [inline data/assets/scene/solarsystem/missions/perseverance/mission.asset:1]: @TODONE

File: data/assets/scene/solarsystem/missions/rosetta/mission.asset [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+local Mission = {
+  Name = "Rosetta",
+  TimeRange = { Start = "2004 MAR 02 00:00:00", End = "2016 SEP 30 00:00:00" },
+  Description = "ESA's Rosetta mission was the first to rendezvous with a comet, the first to follow a comet on its orbit around the Sun, and the first to deploy a lander to a comet's surface. Comets are time capsules containing primitive material left over from the epoch when the Sun and its planets formed. By studying the gas, dust and structure of the nucleus and organic materials associated with the comet, via both remote and in situ observations, the Rosetta mission is unlocking the history and evolution of our Solar System.",
+  Media = { Image = "https://www.cosmos.esa.int/documents/522118/522182/Rosetta_logo.png/cd85878d-2fac-e086-e32a-df0eaee5e505?t=1614026031720" },
+  Milestones = {
+    {
+      Name = "Launch",
+      Date = 	"2004 MAR 02 09:00:00"
```
[the remark this answers is not in the document]
Comment (alexanderbock) [reply data/assets/scene/solarsystem/missions/rosetta/mission.asset:11]: same for a bunch of others below

File: data/profiles/voyager.profile [hunk 9]
```diff
@@ -68,9 +69,9 @@
     }
   ],
   "mark_nodes": [
+    "Voyager_1",
+    "Voyager_2",
```
Comment (alexanderbock) [inline data/profiles/voyager.profile:73]: Good catch!

File: include/openspace/mission/mission.h [hunk 10]
```diff
@@ -81,13 +90,43 @@ class MissionPhase {
      */
     const std::string& description() const;
 
+    /**
+     * Returns the associated image of this MissionPhase. If no image is associated, this
+     * string will be empty.
+     * \return The associated image of the MissionPhase or the empty string
```
Comment (alexanderbock) [inline include/openspace/mission/mission.h:96]: We'd have an extra empty
` *     `
line in here to separate the description from the return values

File: modules/server/include/topics/missiontopic.h [hunk 11]
```diff
@@ -0,0 +1,51 @@
+/*****************************************************************************************
+ *                                                                                       *
+ * OpenSpace                                                                             *
+ *                                                                                       *
+ * Copyright (c) 2014-2021                                                               *
```
Comment (alexanderbock) [inline modules/server/include/topics/missiontopic.h:5]: -> 2023

File: modules/server/include/topics/missiontopic.h [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,   *
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A         *
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#ifndef __OPENSPACE_MODULE_SERVER___MISSION_TOPIC___H__
+#define __OPENSPACE_MODULE_SERVER___MISSION_TOPIC___H__
```
Comment (alexanderbock) [inline modules/server/include/topics/missiontopic.h:26]: I'd probably shorten this to 
`#ifndef __OPENSPACE_MODULE_SERVER___MISSIONTOPIC___H__`
and likewise for the #define and the #endif in the bottom

File: modules/server/include/topics/missiontopic.h [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+#include <openspace/mission/mission.h>
+
+#include <modules/server/include/topics/topic.h>
```
Comment (alexanderbock) [inline modules/server/include/topics/missiontopic.h:30]: These two includes would be reversed since the Topic header is first as it is the base class that we are deriving from

File: modules/server/src/topics/missiontopic.cpp [hunk 14]
```diff
@@ -0,0 +1,124 @@
+/*****************************************************************************************
+ *                                                                                       *
+ * OpenSpace                                                                             *
+ *                                                                                       *
+ * Copyright (c) 2014-2021                                                               *
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:5]: -> 2023

File: modules/server/src/topics/missiontopic.cpp [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+#include <modules/server/include/topics/missiontopic.h>
+
+#include <modules/server/include/connection.h>
+#include <modules/server/include/jsonconverters.h>
+#include <openspace/engine/globals.h>
+#include <openspace/mission/missionmanager.h>
+#include <openspace/util/time.h>
+#include <ghoul/logging/logmanager.h>
+#include <modules/spacecraftinstruments/util/imagesequencer.h>
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:33]: This would come higher up together with the other modules

File: modules/server/src/topics/missiontopic.cpp [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+#include <openspace/util/spicemanager.h>
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:34]: and this together with the other openspace includes

File: modules/server/src/topics/missiontopic.cpp [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+nlohmann::json MissionTopic::missionJson() const {
+    const std::map<std::string, Mission>& missions =
+        global::missionManager->missionMap();
+
+    ImageSequencer& sequencer = ImageSequencer::ref();
+    const std::vector<double>& captureTimes = sequencer.captureProgression();
+    std::vector<std::string> captureTimesString(captureTimes.size());
+
+    for (int i = 0; i < captureTimes.size(); i++) {
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:52]: should be `size_t i` since `.size()` returns that type

File: modules/server/src/topics/missiontopic.cpp [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+        const std::string& str = SpiceManager::ref().dateFromEphemerisTime(
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:53]: Better to do: `std::string str = Spice...`  and then `captureTimesString[i] = std::move(str)` since that is what is happening already, just less visibly so

File: modules/server/src/topics/missiontopic.cpp [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+            sequencer.nextCaptureTime(captureTimes[i]),
+            "YYYY-MM-DDTHR:MN:SC"
+        );
+        captureTimesString[i] = str;
+    }
+    json json;
+    for (auto const& [name, mission] : missions) {
+        nlohmann::json missionJson = createPhaseJson(mission);
+        missionJson["capturetimes"] = captureTimesString;
+        json.push_back(missionJson);
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:63]: also `json.push_back(std::move(missionJson))`

File: modules/server/src/topics/missiontopic.cpp [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    return json;
+}
+
+nlohmann::json MissionTopic::createPhaseJson(const MissionPhase& phase) const {
+
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:70]: Extra empty line

File: modules/server/src/topics/missiontopic.cpp [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+    for (const MissionPhase& missionPhase : phase.phases()) {
+        json subphaseJson = createPhaseJson(missionPhase);
+        phases.push_back(subphaseJson);
+    }
+    
+    json milestones = json::array();
+    const std::vector<Milestone>& dates = phase.milestones();
+    for (const Milestone& date : dates) {
+        json jsonDate = {
+            { "date", std::string(date.date.ISO8601())},
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:81]: Missing space between closing `)` and `}`

File: modules/server/src/topics/missiontopic.cpp [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+            { "name", date.name }
+        };
+        
+        if (date.description.has_value()) {
+            jsonDate["description"] = date.description.value();
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:86]: Quicker way to get to the value of an optional:
`jsonDate["description"] = *date.description;`

File: modules/server/src/topics/missiontopic.cpp [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        if (date.image.has_value()) {
+            jsonDate["image"] = date.image.value();
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:89]: same here

File: modules/server/src/topics/missiontopic.cpp [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        if (date.link.has_value()) {
+            jsonDate["link"] = date.link.value();
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:92]: same here

File: modules/server/src/topics/missiontopic.cpp [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        milestones.push_back(jsonDate);
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:94]: -> `std::move(jsonDate)`

File: modules/server/src/topics/missiontopic.cpp [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    std::string startTimeString = std::string(Time(phase.timeRange().start).ISO8601());
+    std::string endTimeString = std::string(Time(phase.timeRange().end).ISO8601());
+
+    nlohmann::json phaseJson = {
+        { "name", phase.name() },
+        { "description", phase.description() },
+        { "actions" , phase.actions() },
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:103]: Extra space before the `,`

File: modules/server/src/topics/missiontopic.cpp [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+        { "timerange", {
+            { "start" ,startTimeString },
+            { "end" ,endTimeString }
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:106]: the space is on the wrong side of the `,`

File: modules/server/src/topics/missiontopic.cpp [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+        }},
+        { "phases", phases },
+        { "media",{
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:109]: Missing space after the `,`

File: modules/server/src/topics/missiontopic.cpp [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+            { "image", phase.image() },
+            { "link", phase.link() }
+        }},
+        { "milestones" , milestones }
```
Comment (alexanderbock) [inline modules/server/src/topics/missiontopic.cpp:113]: Extra space before the `,`

File: src/mission/mission.cpp [hunk 30]
```diff
@@ -114,6 +157,29 @@ MissionPhase::MissionPhase(const ghoul::Dictionary& dictionary) {
             ));
         }
     }
+
+    if (p.actions.has_value()) {
```
Comment (alexanderbock) [inline src/mission/mission.cpp:161]: The `if` shouldn't be necessary since you are already using `value_or`

File: src/mission/mission.cpp [hunk 31]
```diff
@@ -114,6 +157,29 @@ MissionPhase::MissionPhase(const ghoul::Dictionary& dictionary) {
...
+                newDate.description = p.milestones.value()[i].description.value();
+            }
+            if (p.milestones.value()[i].image.has_value()) {
+                newDate.image = p.milestones.value()[i].image.value();
+            }
+            if (p.milestones.value()[i].link.has_value()) {
+                newDate.link = p.milestones.value()[i].link.value();
+            }
+            _milestones.emplace_back(newDate);
+        }
```
Comment (alexanderbock) [inline src/mission/mission.cpp:181]: Since you don't use the `i` parameter anywhere else, how about:
```suggestion
        for (const Milestone& milestone : *p.milestones) {
            std::string name = milestone.name;
            Time newTime = Time(milestone.date);
            Milestone newDate = { name, newTime };
            if (milestone.description.has_value()) {
                newDate.description = milestone.description.value();
            }
            if (milestone.image.has_value()) {
                newDate.image = milestone.image.value();
            }
            if (milestone.link.has_value()) {
                newDate.link = milestone.link.value();
            }
            _milestones.emplace_back(newDate);
        }
```

Comment (alexanderbock) [reply src/mission/mission.cpp:181]: And then one simplification more:
```cpp
newDate.description = milestone.description.value_or(newDate.description);
newDate.image = milestone.image.value_or(newDate.image);
newDate.link = milestone.link.value_or(newDate.link);
```

File: src/mission/missionmanager.cpp [hunk 32]
```diff
@@ -105,6 +97,11 @@ const Mission& MissionManager::currentMission() {
     return _currentMission->second;
 }
 
+const std::map<std::string, Mission>& MissionManager::missionMap() {
+    return _missionMap;
+}
+
+
```
Comment (alexanderbock) [inline src/mission/missionmanager.cpp:104]: Extra empty line

File: data/assets/scene/solarsystem/missions/osirisrex/actions.asset (hunk 2 above)
Comment (ylvaselling) [reply data/assets/scene/solarsystem/missions/osirisrex/actions.asset:95]: I just copied things from the script log - but yeah, no we don't

File: src/mission/mission.cpp (hunk 31 above)
Comment (ylvaselling) [reply src/mission/mission.cpp:181]: I can't convert to Milestone from parameters - `no suitable user-defined conversion exists`

Comment (alexanderbock) [reply src/mission/mission.cpp]: Sorry; should be `Parameters::Milestone` as the type for the iterator variable, of course

**Comment by ylvaselling** [comment]
> I'm sure it has been there before, but only seeing it now; when the mouse cursor is over the Mission panel, the bottom bar also highlights: image vs image
> 
> _super_ minor detail of course

This is how all our popups are designed and nothing specific to this PR - I'll make an issue
